prov/rstream/src/rstream_msg.c

```diff
@@ -252,20 +242,22 @@ static ssize_t rstream_send_ctrl_msg(struct rstream_ep *ep, uint32_t cq_data)
 		msg.data = cq_data;
 
 		ret = fi_sendmsg(ep->ep_fd, &msg, FI_REMOTE_CQ_DATA);
-		if (ret != 0)
+		if (ret != 0) {
+			fastlock_release(&ep->tx_data_lock);
 			return ret;
+		}
 
 		if (ep->qp_win.tx_credits > 0) {
 			ep->qp_win.tx_credits--;
 		} else {
 			ep->qp_win.ctrl_credits--;
-			rstream_process_cq(ep, RSTREAM_TX_MSG_COMP);
 		}
 	}
 
 	assert(ep->qp_win.target_rx_credits > 0);
 	ep->qp_win.target_rx_credits--;
 
+	fastlock_release(&ep->tx_data_lock);
 	return ret;
```

Comment:
Now that locks have been added, it would be better to insert a label here and use a goto to jump to the label on any abort paths.